### PR TITLE
Create model index column

### PR DESCRIPTION
**Pull request by juliasilge**
Here is some initial work on a model index column for tune, addressing #87 and #209. I wanted you to take a look @topepo before I went much further.

This creates a model index column in `.metrics`:

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6          ✓ recipes   0.1.12    
#> ✓ dials     0.0.7          ✓ rsample   0.0.7     
#> ✓ dplyr     1.0.0          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.1          ✓ tune      0.1.0.9000
#> ✓ infer     0.5.1          ✓ workflows 0.1.1     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.6.9000
#> ✓ purrr     0.3.4
#> ── Conflicts ──────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()

set.seed(6735)
folds <- vfold_cv(mtcars, v = 5)

car_rec <-
  recipe(mpg ~ ., data = mtcars) %>%
  step_normalize(all_predictors())

svm_mod <-
  svm_rbf(cost = tune(), rbf_sigma = tune()) %>%
  set_engine("kernlab") %>%
  set_mode("regression")

# Use a space-filling design with 7 points
set.seed(3254)
svm_res <- tune_grid(svm_mod, car_rec, resamples = folds, grid = 7)

collect_metrics(svm_res)
#> # A tibble: 14 x 8
#>        cost rbf_sigma .model .metric .estimator  mean     n std_err
#>       <dbl>     <dbl> <chr>  <chr>   <chr>      <dbl> <int>   <dbl>
#>  1  0.00247  9.31e- 3 Model5 rmse    standard   5.94      5  0.966 
#>  2  0.00247  9.31e- 3 Model5 rsq     standard   0.765     5  0.0589
#>  3  0.0126   2.39e- 6 Model6 rmse    standard   5.96      5  0.970 
#>  4  0.0126   2.39e- 6 Model6 rsq     standard   0.770     5  0.0590
#>  5  0.0198   1.96e-10 Model7 rmse    standard   5.96      5  0.970 
#>  6  0.0198   1.96e-10 Model7 rsq     standard   0.772     5  0.0579
#>  7  0.304    1.17e- 1 Model4 rmse    standard   3.91      5  0.652 
#>  8  0.304    1.17e- 1 Model4 rsq     standard   0.679     5  0.0595
#>  9  0.459    4.59e- 8 Model2 rmse    standard   5.96      5  0.970 
#> 10  0.459    4.59e- 8 Model2 rsq     standard   0.771     5  0.0591
#> 11  4.53     4.20e- 4 Model3 rmse    standard   4.13      5  0.741 
#> 12  4.53     4.20e- 4 Model3 rsq     standard   0.771     5  0.0588
#> 13 23.2      6.84e- 7 Model1 rmse    standard   5.94      5  0.967 
#> 14 23.2      6.84e- 7 Model1 rsq     standard   0.771     5  0.0591
```

<sup>Created on 2020-06-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Or it says "Recipe1", etc, if only the recipe was tuned:

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6          ✓ recipes   0.1.12    
#> ✓ dials     0.0.7          ✓ rsample   0.0.7     
#> ✓ dplyr     1.0.0          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.1          ✓ tune      0.1.0.9000
#> ✓ infer     0.5.1          ✓ workflows 0.1.1     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.6.9000
#> ✓ purrr     0.3.4
#> ── Conflicts ──────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()

set.seed(6735)
folds <- vfold_cv(mtcars, v = 5)

spline_rec <-
  recipe(mpg ~ ., data = mtcars) %>%
  step_ns(disp, deg_free = tune("disp")) %>%
  step_ns(wt, deg_free = tune("wt"))

lin_mod <-
  linear_reg() %>%
  set_engine("lm")

# manually create a grid
spline_grid <- expand.grid(disp = 2:5, wt = 2:5)

# Warnings will occur from making spline terms on the holdout data that are
# extrapolations.
spline_res <-
  tune_grid(lin_mod, spline_rec, resamples = folds, grid = spline_grid)
collect_metrics(spline_res)
#> # A tibble: 32 x 8
#>     disp    wt .model   .metric .estimator  mean     n std_err
#>    <int> <int> <chr>    <chr>   <chr>      <dbl> <int>   <dbl>
#>  1     2     2 Recipe1  rmse    standard   3.11      5   0.168
#>  2     2     2 Recipe1  rsq     standard   0.651     5   0.135
#>  3     2     3 Recipe5  rmse    standard   3.04      5   0.216
#>  4     2     3 Recipe5  rsq     standard   0.677     5   0.169
#>  5     2     4 Recipe9  rmse    standard   3.53      5   0.358
#>  6     2     4 Recipe9  rsq     standard   0.609     5   0.166
#>  7     2     5 Recipe13 rmse    standard   3.99      5   0.484
#>  8     2     5 Recipe13 rsq     standard   0.527     5   0.155
#>  9     3     2 Recipe2  rmse    standard   2.54      5   0.207
#> 10     3     2 Recipe2  rsq     standard   0.711     5   0.161
#> # … with 22 more rows
```

<sup>Created on 2020-06-11 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Still lots to do here, obviously, but take a look.

### CODE REVIEW

**Comment by topepo** [comment]
Should we use `.config`  instead of `.model`? We have other methods for figuring out where the parameter originated.

Also, as a sacrifice to my OCD, can you use `recipes::names0()` to make the values? 

```r
> recipes::names0(10, "config")
 [1] "config01" "config02" "config03" "config04" "config05" "config06"
 [7] "config07" "config08" "config09" "config10"
```

There are some places where this column will have to be added to the list of exclusions: `get_param_columns()`, `prob_summarize()`, `collector()`, and perhaps others. Maybe we should probably keep an internal vector of extra columns to keep things centralized.

**Comment by juliasilge** [comment]
That is looking pretty good:

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6          ✓ recipes   0.1.12    
#> ✓ dials     0.0.7          ✓ rsample   0.0.7     
#> ✓ dplyr     1.0.0          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.1          ✓ tune      0.1.0.9000
#> ✓ infer     0.5.1          ✓ workflows 0.1.1     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.6.9000
#> ✓ purrr     0.3.4
#> ── Conflicts ──────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()

set.seed(6735)
folds <- vfold_cv(mtcars, v = 5)

car_rec <-
  recipe(mpg ~ ., data = mtcars) %>%
  step_normalize(all_predictors())

svm_mod <-
  svm_rbf(cost = tune(), rbf_sigma = tune()) %>%
  set_engine("kernlab") %>%
  set_mode("regression")

# Use a space-filling design with 7 points
set.seed(3254)
svm_res <- tune_grid(svm_mod, car_rec, resamples = folds, grid = 7)

collect_metrics(svm_res)
#> # A tibble: 14 x 8
#>    .config     cost rbf_sigma .metric .estimator  mean     n std_err
#>    <chr>      <dbl>     <dbl> <chr>   <chr>      <dbl> <int>   <dbl>
#>  1 Model1  23.2      6.84e- 7 rmse    standard   5.94      5  0.967 
#>  2 Model1  23.2      6.84e- 7 rsq     standard   0.771     5  0.0591
#>  3 Model2   0.459    4.59e- 8 rmse    standard   5.96      5  0.970 
#>  4 Model2   0.459    4.59e- 8 rsq     standard   0.771     5  0.0591
#>  5 Model3   4.53     4.20e- 4 rmse    standard   4.13      5  0.741 
#>  6 Model3   4.53     4.20e- 4 rsq     standard   0.771     5  0.0588
#>  7 Model4   0.304    1.17e- 1 rmse    standard   3.91      5  0.652 
#>  8 Model4   0.304    1.17e- 1 rsq     standard   0.679     5  0.0595
#>  9 Model5   0.00247  9.31e- 3 rmse    standard   5.94      5  0.966 
#> 10 Model5   0.00247  9.31e- 3 rsq     standard   0.765     5  0.0589
#> 11 Model6   0.0126   2.39e- 6 rmse    standard   5.96      5  0.970 
#> 12 Model6   0.0126   2.39e- 6 rsq     standard   0.770     5  0.0590
#> 13 Model7   0.0198   1.96e-10 rmse    standard   5.96      5  0.970 
#> 14 Model7   0.0198   1.96e-10 rsq     standard   0.772     5  0.0579
```

<sup>Created on 2020-06-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by juliasilge** [comment]
Do we hate or feel OK with this kind of ordering result in `collect_metrics()`? The `.config` column is in the original order ("Recipe1", "Recipe2", etc) in the `.metrics` column.

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6          ✓ recipes   0.1.12    
#> ✓ dials     0.0.7          ✓ rsample   0.0.7     
#> ✓ dplyr     1.0.0          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.1          ✓ tune      0.1.0.9000
#> ✓ infer     0.5.1          ✓ workflows 0.1.1     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.6.9000
#> ✓ purrr     0.3.4
#> ── Conflicts ──────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()

set.seed(6735)
folds <- vfold_cv(mtcars, v = 5)

spline_rec <-
  recipe(mpg ~ ., data = mtcars) %>%
  step_ns(disp, deg_free = tune("disp")) %>%
  step_ns(wt, deg_free = tune("wt"))

lin_mod <-
  linear_reg() %>%
  set_engine("lm")

# manually create a grid
spline_grid <- expand.grid(disp = 2:5, wt = 2:5)

# Warnings will occur from making spline terms on the holdout data that are
# extrapolations.
spline_res <-
  tune_grid(lin_mod, spline_rec, resamples = folds, grid = spline_grid)
collect_metrics(spline_res)
#> # A tibble: 32 x 8
#>    .config   disp    wt .metric .estimator  mean     n std_err
#>    <chr>    <int> <int> <chr>   <chr>      <dbl> <int>   <dbl>
#>  1 Recipe1      2     2 rmse    standard   3.11      5   0.168
#>  2 Recipe1      2     2 rsq     standard   0.651     5   0.135
#>  3 Recipe10     3     4 rmse    standard   3.24      5   0.352
#>  4 Recipe10     3     4 rsq     standard   0.637     5   0.158
#>  5 Recipe11     4     4 rmse    standard   3.01      5   0.475
#>  6 Recipe11     4     4 rsq     standard   0.648     5   0.151
#>  7 Recipe12     5     4 rmse    standard   3.30      5   0.397
#>  8 Recipe12     5     4 rsq     standard   0.611     5   0.153
#>  9 Recipe13     2     5 rmse    standard   3.99      5   0.484
#> 10 Recipe13     2     5 rsq     standard   0.527     5   0.155
#> # … with 22 more rows
```

<sup>Created on 2020-06-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by topepo** [comment]
The row ordering is fine with me. 

Can you put the column all the way to the right though? That should be the least likely to be shown when the column width is small.

**Comment by topepo** [comment]
I don't think that we need to differentiate values of "model" and "recipe". I think that "config" would be better (but open to suggestions). When we add post-process or other stages, we have to account for new values. Would we ever process the values to differentiate recipes and models?

**Comment by juliasilge** [comment]
It could be helpful to be able to ascertain from `.config` which results are from the sample recipe vs. model. I currently have it implemented within the two loops so it comes out like this:

``` r
library(tidymodels)
#> ── Attaching packages ───────────────────────────────────────────────────────────────── tidymodels 0.1.0 ──
#> ✓ broom     0.5.6          ✓ recipes   0.1.12    
#> ✓ dials     0.0.7          ✓ rsample   0.0.7     
#> ✓ dplyr     1.0.0          ✓ tibble    3.0.1     
#> ✓ ggplot2   3.3.1          ✓ tune      0.1.0.9000
#> ✓ infer     0.5.1          ✓ workflows 0.1.1     
#> ✓ parsnip   0.1.1.9000     ✓ yardstick 0.0.6.9000
#> ✓ purrr     0.3.4
#> ── Conflicts ──────────────────────────────────────────────────────────────────── tidymodels_conflicts() ──
#> x purrr::discard() masks scales::discard()
#> x dplyr::filter()  masks stats::filter()
#> x dplyr::lag()     masks stats::lag()
#> x recipes::step()  masks stats::step()

set.seed(6735)
folds <- vfold_cv(mtcars, v = 5)

spline_rec <-
  recipe(mpg ~ ., data = mtcars) %>%
  step_ns(disp, deg_free = tune("disp")) %>%
  step_ns(wt, deg_free = tune("wt"))

lasso_spec <-
  linear_reg(penalty = tune(), mixture = 1) %>%
  set_engine("glmnet")

lasso_wf <-
  workflow() %>%
  add_model(lasso_spec) %>%
  add_recipe(spline_rec)

lasso_res <- tune_grid(lasso_wf, resamples = folds, grid = 7, control = control_grid(save_pred = TRUE))

collect_metrics(lasso_res)
#> # A tibble: 14 x 9
#>     penalty  disp    wt .metric .estimator   mean     n std_err .config       
#>       <dbl> <int> <int> <chr>   <chr>       <dbl> <int>   <dbl> <chr>         
#>  1 1.15e-10     4    11 rmse    standard   11.9       5  5.44   Recipe1_Model1
#>  2 1.15e-10     4    11 rsq     standard    0.462     5  0.173  Recipe1_Model1
#>  3 7.05e- 5    15     7 rmse    standard    6.56      5  3.75   Recipe2_Model1
#>  4 7.05e- 5    15     7 rsq     standard    0.684     5  0.0880 Recipe2_Model1
#>  5 3.82e- 1     1    14 rmse    standard    2.92      5  0.316  Recipe3_Model1
#>  6 3.82e- 1     1    14 rsq     standard    0.673     5  0.0757 Recipe3_Model1
#>  7 4.88e- 6     9     6 rmse    standard    4.80      5  0.669  Recipe4_Model1
#>  8 4.88e- 6     9     6 rsq     standard    0.387     5  0.169  Recipe4_Model1
#>  9 1.67e- 3    11     3 rmse    standard    3.80      5  0.801  Recipe5_Model1
#> 10 1.67e- 3    11     3 rsq     standard    0.598     5  0.151  Recipe5_Model1
#> 11 7.76e- 8     6     2 rmse    standard    2.89      5  0.330  Recipe6_Model1
#> 12 7.76e- 8     6     2 rsq     standard    0.671     5  0.124  Recipe6_Model1
#> 13 1.09e- 8     8    12 rmse    standard    3.13      5  0.533  Recipe7_Model1
#> 14 1.09e- 8     8    12 rsq     standard    0.611     5  0.136  Recipe7_Model1
```

<sup>Created on 2020-06-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by topepo** [comment]
We have that data in a more convenient format in the object's `parameters` attribute. 

But... as I type... could we use your labels in the verbose logging? People have (understandably) mentioned that, during the logging, they don't know what model is being logged. Having the _post hoc_ results match the logged value would be a good feature.

**Comment by juliasilge** [comment]
Yeah, if you [click through to what's changed](https://github.com/tidymodels/tune/pull/227/files#diff-f7284db5f6c6be50eb6d66273d9c0917), check out how these ids are being created at the same time as the logging messages so they match up.

File: R/collect.R [hunk 1]
```diff
@@ -354,5 +354,14 @@ estimate_tune_results <- function(x, ...) {
       std_err = sd(.estimate, na.rm = TRUE)/sqrt(n)
     ) %>%
     dplyr::ungroup()
+
+  if (".config" %in% param_names) {
```
Comment (topepo) [inline R/collect.R:358]: Would it be safer to make this a join instead of column binding?

**Pull request merged**